rt/rt.mjs

```diff
@@ -14,115 +14,164 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-const TAG_SIZE = 3;
-const TAGS = {
-    fixnum: 0,
-    constant: 1,
-    pair: 2,
-    character: 3,
-    string: 4,
-    symbol: 5,
-};
-
-function tag_constant(value, tag) {
-    return (value << TAG_SIZE) | tag;
-}
-
-function replace_tag(ptr, tag) {
-    return tag_constant(extract_value(ptr), tag);
-}
-
-function extract_tag(value) {
-    return value & ((1 << TAG_SIZE) - 1);
-}
-
-function extract_value(value) {
-    return value >>> TAG_SIZE;
-}
-
-const SCHEME_CONSTANTS = {
-    0: false,
-    1: true,
-    2: null,
-    3: "#<eof-object>",
+class SchemeError extends Error {
+    constructor(where, what) {
+        super(`Scheme runtime error in ${where}: ${what}`);
+        this.where = where
+        this.what = what
+    }
 }
 
-const CONSTANTS = {
-    false: tag_constant(0, TAGS.constant),
-    true: tag_constant(1, TAGS.constant),
-    null: tag_constant(2, TAGS.constant),
-    eof: tag_constant(3, TAGS.constant)
+function addToMap(map, key, val) {
+    map.set(key, val);
+    return val;
 }
 
-// Convert a Scheme ptr into a corresponding JavaScript value
-function js_from_scheme(ptr) {
-    switch (extract_tag(ptr)) {
-        case TAGS.fixnum:
-            return ptr >> TAG_SIZE; // sign extending shift so negatives work.
-        case TAGS.constant:
-            return SCHEME_CONSTANTS[extract_value(ptr)];
-        case TAGS.character:
-            return String.fromCharCode(extract_value(ptr));
+export class Pair {
+    constructor(car, cdr) {
+        this.car = car;
+        this.cdr = cdr;
     }
 }
 
-// Convert a JS value into a Scheme ptr
-function scheme_from_js(value) {
-    const type = typeof value;
-    switch(type) {
-        case 'number':
-            return value << TAG_SIZE;
-        case 'boolean':
-            return CONSTANTS[value];
-        default:
-            throw new Error(`Converting a ${type} has not been implemented`);
+export class Closure {
+    constructor(index, nfree) {
+        this.index = index | 0;
+        this.env = [];
     }
 }
 
-function fixnum_from_number(n) {
-    return n << TAG_SIZE;
-}
-
-class SchemeError extends Error {
-    constructor(where, what) {
-        super(`Scheme runtime error in ${where}: ${what}`);
-        this.where = where
-        this.what = what
+export class EOF {};
+const eofValue = new EOF;
+
+function toJS(x, map) {
+    switch (typeof x) {
+    case 'number':
+        if (x & 1)
+            return String.fromCharCode(x >>> 1);
+        return x >> 1;
+    case 'string':
+        if (x.charAt(0) === 's')
+            return x.substring(1);
+        return x.substring(2);
```

Comment:
Why is this `.substring(2)` instead of `1`?

Reply:
Strings get the 's' tag.  Symbols get the 'S' tag.  Symbols end up being tagged with both the string and the symbol tags -- hence substring(2).  Not very well commented!

Reply:
Ah okay, thanks!